### ISSUE TITLE
Add a config option to allow preemptive auth

### ISSUE THREAD

**Issue by igor-kupczynski** [searchbox-io/Jest#207]
Hi Jest Team,

I tried to use the basic auth feature added in 0.1.6
```java
JestClientFactory factory = new JestClientFactory();
factory.setHttpClientConfig(
    new HttpClientConfig.Builder("http://localhost:9200")
        .defaultCredentials("global_user", "global_password")
        .build()
);
```

The problem is that it doesn't do preemptive auth and AFAIK there is no way to enforce it.

Example access log from nginx fronting a test cluster
```
127.0.0.1 - - [26/May/2015:17:27:44 +0200] "POST /index/file/_search HTTP/1.1" 401 195 "-" "Apache-HttpClient/4.4.1 (Java/1.7.0_60)"
127.0.0.1 - user [26/May/2015:17:27:44 +0200] "POST /index/file/_search HTTP/1.1" 200 912 "-" "Apache-HttpClient/4.4.1 (Java/1.7.0_60)"
127.0.0.1 - - [26/May/2015:17:27:44 +0200] "POST /index/file/_search HTTP/1.1" 401 195 "-" "Apache-HttpClient/4.4.1 (Java/1.7.0_60)"
127.0.0.1 - user [26/May/2015:17:27:44 +0200] "POST /index/file/_search HTTP/1.1" 200 132 "-" "Apache-HttpClient/4.4.1 (Java/1.7.0_60)"
127.0.0.1 - - [26/May/2015:17:28:08 +0200] "POST /index/file/d6371446-cbbc-44af-9d3f-e299047fef79/_update HTTP/1.1" 401 195 "-" "Apache-HttpClient/4.4.1 (Java/1.7.0_60)"
127.0.0.1 - user [26/May/2015:17:28:08 +0200] "POST /index/file/d6371446-cbbc-44af-9d3f-e299047fef79/_update HTTP/1.1" 201 96 "-" "Apache-HttpClient/4.4.1 (Java/1.7.0_60)"
127.0.0.1 - - [26/May/2015:17:28:12 +0200] "POST /index/file/d6371446-cbbc-44af-9d3f-e299047fef79/_update HTTP/1.1" 401 195 "-" "Apache-HttpClient/4.4.1 (Java/1.7.0_60)"
127.0.0.1 - user [26/May/2015:17:28:13 +0200] "POST /index/file/d6371446-cbbc-44af-9d3f-e299047fef79/_update HTTP/1.1" 200 96 "-" "Apache-HttpClient/4.4.1 (Java/1.7.0_60)"
```

As you can see, every request starts without sending credentials, it ends up in `401` and then is repeated with basic auth included and ends up as `200`.

This is an http client design decision http://stackoverflow.com/questions/2014700/preemptive-basic-authentication-with-apache-httpclient-4 and an example way to work around this is presented in their svn https://subversion.jfrog.org/jfrog/build-info/trunk/build-info-client/src/main/java/org/jfrog/build/client/PreemptiveHttpClient.java

```java
client.getCredentialsProvider().setCredentials(
                    new AuthScope(AuthScope.ANY_HOST, AuthScope.ANY_PORT),
                    new UsernamePasswordCredentials(userName, password)
            );
            localContext = new BasicHttpContext();

            // Generate BASIC scheme object and stick it to the local execution context
            BasicScheme basicAuth = new BasicScheme();
            localContext.setAttribute("preemptive-auth", basicAuth);

            // Add as the first request interceptor
            client.addRequestInterceptor(new PreemptiveAuth(), 0);
```

The problem is that I see no way to add request interceptor and no way to add this local context on each request.

This is an important feature in context of elasticsearch, as usually you have just one server with one user/pass for your app. So sending each request twice seems really a waste in case of elasticsearch under a heavy load.

Do you agree this is a valid feature to have? Any suggestions on the implementation? I think that from an API consumer point of view the best would be to add new method `preemptiveAuth(boolean)` to `HttpClientConfig.Builder`.

Cheers,
Igor

**Comment by kramer** [comment]
> As of version 4.1 HttpClient automatically caches information about hosts it has successfully authenticated with. Please note that one must use the same execution context to execute logically related requests in order for cached authentication data to propagate from one request to another. Authentication data will be lost as soon as the execution context goes out of scope.

 so it should be enough for us to use a

**Comment by kramer** [comment]
As far as I understood interceptor is only useful when you don't want to or can't set the context at each request execution, but we can easily do that so I don't see the need.

Another safer solution is not blindly doing preemptive auth but using the auth caching mechanism as described in [here](https://hc.apache.org/httpcomponents-client-ga/tutorial/html/authentication.html#d5e703):

> As of version 4.1 HttpClient automatically caches information about hosts it has successfully authenticated with. Please note that one must use the same execution context to execute logically related requests in order for cached authentication data to propagate from one request to another. Authentication data will be lost as soon as the execution context goes out of scope.

This way auth challenge is done only once and all we need to do is give the same context (which we let httpclient to fill in during first request) in all `httpClient.execute` calls. The problem now is context instances are **not thread safe** and we have an async client... Also I'm not sure if it is good practice to just use the same context for the lifetime of the client. Implementation of this would sure include some locks here and there so even though nicer I'm ready to give up on this solution unless someone comes up with a clean way to implement it.

Otherwise yes it's a valid feature, would you like to take it?

**Comment by puszczyk** [comment]
Thanks for the detailed reply.

I'd like to take this feature, but the thing it is that next two weeks I'm on PTO and then probably will have a lot of stuff to catch up. So I can't promise any ETA. If lets say in 1-1.5 months time there is no other person I'll give it a shot.

**Comment by kramer** [comment]
No probs. I don't think we'll have a new version/release in a month so I'll leave this for you (unless I spontaneously end up with a commit).

**Comment by kramer** [comment]
@puszczyk Have you get around to implementing this? If not I may start it.

**Comment by puszczyk** [comment]
Sorry, I do not think I can do it now, so I have to yield the right to implement it ;-)

**Comment by david-tejedor** [comment]
Any updates on this issue?
If not I have a look

**Comment by kramer** [comment]
Please feel free to do so @david-tejedor :)

**Comment by alkiskal** [comment]
Hello all, do you intend to add this in the next release? It is very important for my use case and I believe for a lot of other users too.
